design/oauth2-device-authorization-grant.md

```diff
@@ -0,0 +1,144 @@
+
+This endpoint provides user code verification and user authentication, which we will discuss in User Interaction section.
+
+### Token Endpoint
+
+Token request for the spec is represented with new grant type of `urn:ietf:params:oauth:grant-type:device_code`. We add the process of this new grant type into an existing TokenEndpoint class for OAuth 2.0/OIDC.
+
+
+## User Interaction when verifying a user code
+
+The spec says that the details of the user interaction when processing the verification are up to the authorization server in [3.3. User Interaction](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.3). There are two methods for the process which are imeplemented by major IdPs. *Login First* or *User Code Verification First*.
+
+### Login First
+
+In this case, the end-user need to login first before entering a user code in the user interaction. It means the authorization server must provide a shared authentication flow because the server can't detect the client ID yet which is bind to the user code. However, the risk getting user code brute force attack is low as attackers must log in first.
+
+### User Code Verification First
+
+In this case, the end-user need to enter a user code first in the user interaction. It means the authorization server can provide different authentication flow per client because the server can detect the client ID from the user code. However, the risk getting user code brute force attack is higher than Login First as attackers can try to enter user codes freely.
+
+### Which user interaction does keycloak should support?
+
+User Code Verification First is adopted by major IdPs like Google, Microsoft, and Salesforce. So it seems to be a common practice. Also, it can provide a more flexible authentication flow per device client. Keycloak already provides a feature that a client can use specific browser flow by using the Authentication Flow Overrides option. Therefore, it's suitable for supporting User Code Verification First.
+
+## Consent
+
+The spec says that the authorization server SHOULD display information about the device so that the person can notice if a software client was attempting to impersonating a hardware device. So It should always display consent screen when end-user do log in even if consent required is true or the user already consented.
+
+## User Code Format
+
+The spec says about user code format patterns in [6.1. User Code Recommendations](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-6.1). It's good to be able to provide an SPI for user code format customization. We implement a case-insensitive eight-letter format as a default implementation of the SPI.
```

Comment:
Why not use `WDYG-MFKS` ? It looks better than space, IMO. The code is not subject to doubts whether or not you have spaces in code.

Reply:
What doubts? Space is optional so it doesn't matter. You can type it as [xxxxyyyy] [xxxx yyyy] or even [xxxx] [yyyy].

Google uses spaces, Apple uses hyphens. I find spaces more readable. Further, with optional spaces the input can be split into multiple inputs if wanted and just joined together without having to consider what the delimeter is as it is optional:

[XXXX] [YYYY] - instead of [XXXX-YYYY]

Reply:
When I refer to [Google documents](https://developers.google.com/identity/protocols/OAuth2ForDevices#step-2-handle-the-authorization-server-response), it seems to use dashes as a delimiter.
Also dashes is used in the spec [6.1](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-6.1) as an example such as "WDJB-MJHT". Therefore, I feel that dashes is preferable as default. Also, in my proposal this format is customizable with SPI.

Reply:
OK, let's go with dashes then. I was thinking about application passwords feature that Google has, where it uses optional spaces.